### PR TITLE
#160451498 Validate user details on registration

### PR DESCRIPTION
**Pull request by kampkelly**
#### What does this PR do?
This pull request validates the user details when signing up.

#### Description of Task to be completed?
- Write tests for validating user details
- Write the validations for user details in a middleware
- Move index.js from root to server folder

#### How should this be manually tested?
After cloning the project, run `npm install`, then `npm run watch` to start the server. Using postman on sending incorrect details to the API endpoint to signup, the user will not get signed up and will get a descriptive response message.

#### Any background context you want to provide?
None
#### What are the relevant pivotal tracker stories?
#160451498
#### Questions:```
None

### CODE REVIEW

**Review by onwuvic — COMMENTED** [review]
LTGM

File: server/middlewares/userValidations.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import Joi from 'joi';
+/**
+  * @class UserValidationsController
+  * @description Validate User Input
+  */
+class UserValidations {
+  /**
+   *
```
Comment (s-akinrele) [inline server/middlewares/userValidations.js:8]: add more code comments on this file to explain what you are doing.

File: server/middlewares/userValidations.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  static validateSignup(req, res, next) {
+    req.body.email = req.body.email.toLowerCase().replace(/\s/g, '');
+    req.body.username = req.body.username.replace(/\s/g, '');
+    const joiSchema = {
+      email: Joi.string().email().required(),
+      password: Joi.string().alphanum().min(8).required(),
+      username: Joi.string().min(2).required(),
+    };
+    const joiOptions = { abortEarly: false };
+    const {
```
Comment (s-akinrele) [inline server/middlewares/userValidations.js]: put this on one line.

File: server/middlewares/userValidations.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      username, email, password
+    } = req.body;
+    Joi.validate({
```
Comment (s-akinrele) [inline server/middlewares/userValidations.js]: add space one line space above

Comment (s-akinrele) [reply server/middlewares/userValidations.js]:
```
Joi.validate({email, username, password} , joiSchema, joiOptions, (err) => { 
```

File: server/middlewares/userValidations.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      email, username, password,
+    }, joiSchema, joiOptions, (err) => {
+      if (err) {
+        const getErrorMessages = error => error.message.replace(/"/g, '');
```
Comment (s-akinrele) [inline server/middlewares/userValidations.js:34]: add code comment here

**Review by s-akinrele — CHANGES_REQUESTED** [review]
I had some comments

File: server/routes/index.js [hunk 5]
```diff
@@ -3,6 +3,6 @@ import api from './api';
 
 const router = express.Router();
 
-router.use('/api', api);
+router.use('/api/v1', api);
```
Comment (kampkelly) [inline server/routes/index.js:6]: this specifies that this is version one of the api. So all the endpoints will be prefixed with `api/v1`.

File: server/middlewares/userValidations.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import Joi from 'joi';
```
Comment (kampkelly) [inline server/middlewares/userValidations.js:1]: this imports Joi which is an npm package. This package can be used to validate user inputs to ensure that only the correct format and datatypes are being sent to the model to be saved.

**Review by s-akinrele — APPROVED** [review]
GTM. thanks for the updates.

**Pull request merged**